### PR TITLE
Parse comment before sending it to the back-end to preserve HTML entities

### PR DESCRIPTION
**Pull request by pecanoro**
### Details
WAF strips all HTML tags due to security reasons so if we don't encode the message, we lose them when showing them back to the user.

### Fixed Issues

Part of https://github.com/Expensify/App/issues/15131

### Tests
1. Send/request money to any account. 
2. In the `what's for` section, enter `<a> blabla &nbsp </a>`
3. Send the request, it should not fail and the message in the request in the chat should look like this:

image

4. Repeat the process but this time sending money, the message should be displayed exactly the same without missing any characters.
5. Click on the details for the request, the message should be properly formatted as well:

image

- [x] Verify that no errors appear in the JS console

### Offline tests
1. Send/request money to any account. 
2. In the `what's for` section, enter `<a> blabla &nbsp </a>`
3. Send the request, it should not fail and the message in the request in the chat should look like this:

image

4. Repeat the process but this time sending money, the message should be displayed exactly the same without missing any characters.
5. Click on the details for the request, the message should be properly formatted as well:

image

### QA Steps
1. Send/request money to any account. 
2. In the `what's for` section, enter `<a> blabla &nbsp </a>`
3. Send the request, it should not fail and the message in the request in the chat should look like this:

image

4. Repeat the process but this time sending money, the message should be displayed exactly the same without missing any characters.
5. Click on the details for the request, the message should be properly formatted as well:

image

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

image

</details>

<details>
<summary>Mobile Web - Chrome</summary>

image

</details>

<details>
<summary>Mobile Web - Safari</summary>

image

</details>

<details>
<summary>Desktop</summary>

image

</details>

<details>
<summary>iOS</summary>

image

</details>

<details>
<summary>Android</summary>

image

</details>

### CODE REVIEW

**Comment by tienifr** [comment]
@pecanoro I have checked the changes. It seems that you haven't addressed issues with the message in BrowserNotification, LHN. Do I have to fix them in [my PR](https://github.com/Expensify/App/pull/15340)?

**Comment by pecanoro** [comment]
Removing HOLD

**Comment by sobitneupane** [comment]
@pecanoro 
The issue still exists in IOU Message in IOU Details.
Expected Result: `Requested NPR 70.00 from &#38 for <a> blabla &nbsp </a>`
Screenshot 2023-03-21 at 15 12 15

**Comment by pecanoro** [comment]
@sobitneupane I am confused, which part do you say it's not working yet? I didn't change normal messages, only IOU messages. Do you mean that the IOU browser notification?

**Comment by sobitneupane** [comment]
@pecanoro https://github.com/Expensify/App/pull/16055#issuecomment-1477525578 is a pre-existing issue and is not resulted by this PR. The issue exists both for simple message and IOU message.

**Comment by pecanoro** [comment]
Gotcha! I think that issue is being handled here https://github.com/Expensify/App/issues/15508

**Comment by sobitneupane** [comment]
Okay Thanks. Then we have one issue left. https://github.com/Expensify/App/pull/16055#issuecomment-1477524781

**Comment by pecanoro** [comment]
However, I actually think the converted characters should be handled by https://github.com/Expensify/App/issues/15508

**Comment by sobitneupane** [comment]
I raised it here because IOU Message is displayed correctly in chat but not in IOU Details.

I think we should remove Str.htmlDecode and replace html with text in the following line.
https://github.com/Expensify/App/blob/3be4d99c59ec18dcf8511844798b75c9818dd759/src/components/ReportTransaction.js#L80

We can either handle it here or in the other PR from @tienifr.

**Comment by sobitneupane** [comment]
I have added comment in https://github.com/Expensify/App/issues/15508#issuecomment-1477698557 to make sure https://github.com/Expensify/App/pull/16055#issuecomment-1477525578 issue gets resolved there.

**Comment by pecanoro** [comment]
> I raised it here because IOU Message is displayed correctly in chat but not in IOU Details.

What do you mean by that? I tested that and it was showing correctly, I even added screenshots. Sorry, I am not sure if I am missing some specific case I didn't test. Did you use staging for the changes?

**Comment by pecanoro** [comment]
@sobitneupane Or are you are taking about the profile name? I didn't fix that, only focused on the message.

**Comment by sobitneupane** [comment]
@pecanoro 
In the following screenshot, IOU Message (inside green box) is the expected IOU Message. You can notice that IOU Message (inside red box) has "&" instead of "&#38".
Screenshot 2023-03-21 at 19 35 46

**Comment by pecanoro** [comment]
We mentioned that profile names were being handled in https://github.com/Expensify/App/pull/15340.

**Comment by pecanoro** [comment]
@sobitneupane They probably need to be sent encoded to the back-end so they get decoded before being sent back.

File: src/libs/actions/IOU.js [hunk 1]
```diff
@@ -193,11 +193,12 @@ function requestMoney(report, amount, currency, recipientEmail, participant, com
         reportActionsFailureData,
     ];
 
+    const commentText = ReportUtils.getParsedComment(comment);
```
Comment (luacmartins) [inline src/libs/actions/IOU.js:196]: NAB
```suggestion
    const parsedComment = ReportUtils.getParsedComment(comment);
```

**Review by luacmartins — CHANGES_REQUESTED** [review]
I think we need to update the [Split Bill](https://github.com/Expensify/App/blob/aa2e559159aee2e2026c2a8433c2a63a0d9ee7a6/src/libs/actions/IOU.js#L493) flow as well

**Comment by sobitneupane** [comment]
@pecanoro Looks like we have merge conflicts.

**Comment by pecanoro** [comment]
I will update this today, I was on vacation for a few days!

**Comment by sobitneupane** [comment]
### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/25876548/228497801-e30b46d2-7976-416e-89df-c4386381c7b7.mov

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/25876548/228499391-e8e39eef-6cf3-404a-927f-0d34f926229f.mov

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/25876548/228500033-f3dc08e1-3823-4cf9-83c0-c7d94383a8f3.mov

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/25876548/228498825-de3e4b50-d867-46b7-b6dc-b8fc27d74673.mov

</details>

<details>
<summary>iOS</summary>



</details>

<details>
<summary>Android</summary>


</details>

**Review by sobitneupane — APPROVED** [review]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

**Review by luacmartins — APPROVED** [review]

**Pull request merged**